enhancements/ignition-spec-dual-support.md

```diff
@@ -0,0 +1,448 @@
+that a cluster only has one “desiredConfig” which will be translated to spec 3
+when the MCO with dual support detects that the existing configuration of a
+machine is on spec 2 (will happen only once for all existing and new nodes,
+when the MCO with dual support is first deployed onto the cluster). This will
+only be required as part of the MCO.
+
+Note that there exists three types of spec 2 configs:
+ - Those that are directly translatable to spec 3. This is the case for all existing IPI configs.
+ - Those that we cannot translate with certainty, and requires user input for us to correctly translate
+
+During phase 1 we should attempt to translate the existing config to detect
+untranslatable configs. If the cluster can be directly translated to spec 3,
+no action will be required. If not, we will need to alert the user that there
+are untranslatable configs.
+
+During phase two, we should fail updates unless the cluster is fully on spec
+3 config. This effectively means that UPI clusters are at risk when upgrading
+to an MCO with dual support. Mitigation methods are discussed below.
+
+For backwards support (spec 3 to spec 2) the configs are 100% translatable.
+
+
+### RHCOS
+
+Phase 1:
+
+No changes needed.
+
+Phase 2:
+
+The RHCOS bootimage needs to be updated to Ignition package v2.0+ . Required
+dependencies are discussed below.
+
+
+### Installer
+
+Phase 1:
+
+No changes needed.
+
+Phase 2:
+
+The bootstrap Ignition configs are updated to be spec 3. The stub Ignition
+configs for master/worker nodes are updated to spec 3, and referece the correct
+endpoint of the MCS which will serve Ignition spec 3 configs. RHCOS images pinned
+by the installer will be updated in conjuction to ones with Ignition v2 (spec3).
+All spec 2 references are stripped from the installer.
+
+
+### MCO
+
+The MCO and its subcomponents are the most affected by this change. The
+aforementioned spec translator will be housed in the MCO. This means that the
+MCO would need to simultaneously vendor both Ignition v0 and v2, and translate
+existing between the spec versions as needed.
+
+Phase 1:
+
+The MCD has the capability to understand both spec 2 and spec 3 configs,
+and lay down files as needed.
+
+The MCC has the capability to translate spec 2 to spec 3 configs. If the
+translation is completely successful, it will create final rendered spec 3
+configs, and the MCD will be instructed to use the spec 3 configs to lay down
+necessary files. Otherwise the MCD will continue using the existing
+spec 2 configs, and apply new spec 2/spec 3 configs.
+
+The MCS served configs generated by the MCC will be downtranslated by the
+MCC to spec 2. The MCS is not updated to serve spec 3 configs yet.
+
+Phase 2:
+
+All machineconfigs are switched to spec 3. Spec 2 support will exist in the form
+of served Ignition configs to new nodes joining the cluster that have Ignition
+v0 in the bootimage. MCS will handle this with 2 endpoints.
```

Comment:
> MCS will handle this with 2 endpoints.

This should be a single endpoint. Ignition already uses the Accept HTTP header to specify which configs it can accept and the order of preference.

Reply:
cc @yuqi-zhang this is related also to the ignition endpoint contained in that secret created by the installer (and unmanaged)

Reply:
I can update the proposal. We would still first need an option to modify the secret anyways.